subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationRole.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+/**
+ * Default implementation of {@link ConfigurationRole}.
+ */
+public class DefaultConfigurationRole implements ConfigurationRole {
```

Comment:
For the equals method, do you think the name should be included? I would say no, since the name does not affect how the role functions, just how it is displayed, but I could be convinced otherwise.

Reply:
I think it boils down to: when would you have 2 roles with the same usage but different names?  Perhaps when copying a configuration, which already has issues.  If you're using the public references to the pre-built instances it's a moot point, since they are `==`.  It's not mutable, but it does seem to be part of the identity, since it is specified in the constructor.  If you have 2 roles with the same usage and different names you'd want to be able to put them both in a set though, and not worry about only the first one ending up in there.

I suppose I lean slightly towards including the name in `equals` and adding a `hasSameUsage(ConfigurationRole)` method if we ever need to categorize without the role.